sources/imdsclient/src/lib.rs

```diff
@@ -0,0 +1,726 @@
+        // Returns a list of available public keys as '0=my-public-key'
+        let public_key_list = match self.fetch_string("meta-data/public-keys").await {
+            Err(error::Error::NotFound { uri: _ }) => {
+                // this is OK, it just means there are no keys
+                debug!("no available public keys");
+                return Ok(Vec::new());
+            }
+            Err(e) => {
+                return Err(e);
+            }
+            Ok(value) => value,
+        };
+
+        debug!("available public keys '{}'", &public_key_list);
+        info!("Generating targets to fetch text of available public keys");
+        let public_key_targets = build_public_key_targets(&public_key_list);
+
+        let mut public_keys = Vec::new();
+        let target_count: u32 = 0;
+        for target in &public_key_targets {
+            let target_count = target_count + 1;
+            info!(
+                "Fetching public key ({}/{})",
+                target_count,
+                &public_key_targets.len()
+            );
+
+            let public_key_text = self.fetch_string(&target).await?;
+            let public_key = public_key_text.trim_end();
+            // Simple check to see if the text is probably an ssh key.
+            if public_key.starts_with("ssh") {
+                debug!("{}", &public_key);
+                public_keys.push(public_key.to_string())
+            } else {
+                warn!(
+                    "'{}' does not appear to be a valid key. Skipping...",
+                    &public_key
+                );
+                continue;
+            }
+        }
+        if public_keys.is_empty() {
+            warn!("No valid keys found");
+        }
+        Ok(public_keys)
+    }
+
+    /// Helper to fetch bytes from IMDS using the pinned schema version.
+    async fn fetch_bytes<S>(&mut self, end_target: S) -> Result<Vec<u8>>
+    where
+        S: AsRef<str>,
+    {
+        self.fetch_imds(PINNED_SCHEMA, end_target.as_ref()).await
+    }
+
+    /// Helper to fetch a string from IMDS using the pinned schema version.
+    async fn fetch_string<S>(&mut self, end_target: S) -> Result<String>
+    where
+        S: AsRef<str>,
+    {
+        let response_body = self.fetch_imds(PINNED_SCHEMA, end_target).await?;
+        Ok(String::from_utf8(response_body).context(error::NonUtf8Response)?)
+    }
+
+    /// Fetch data from IMDS.
+    async fn fetch_imds<S1, S2>(&mut self, schema_version: S1, target: S2) -> Result<Vec<u8>>
+    where
+        S1: AsRef<str>,
+        S2: AsRef<str>,
+    {
+        let uri = format!(
+            "{}/{}/{}",
+            self.imds_base_uri,
+            schema_version.as_ref(),
+            target.as_ref()
+        );
+        debug!("Requesting {}", &uri);
+        let mut attempt: u8 = 1;
+        let max_attempts: u8 = 3;
+        loop {
+            attempt += 1;
+            ensure!(attempt <= max_attempts, error::FailedFetch { attempt });
```

Comment:
Seems like we'll only try twice?

```suggestion
        let mut attempt: u8 = 0;
        let max_attempts: u8 = 3;
        loop {
            attempt += 1;
            ensure!(attempt <= max_attempts, error::FailedFetch { attempt });
```

Reply:
I think it's correct `attempt <= max_attempts`